sources/src/main/java/com/ullink/slack/simpleslackapi/impl/SlackWebSocketSessionImpl.java

```diff
@@ -334,43 +348,50 @@ private void connectImpl() throws IOException
         establishWebsocketConnection();
     }
 
-    private void establishWebsocketConnection() throws IOException
-    {
+    private void establishWebsocketConnection() throws IOException {
         lastPingSent = 0;
         lastPingAck = 0;
-        WebSocketContainer client = webSocketContainerProvider.getWebSocketContainer();
+        final WebSocketContainer client = webSocketContainerProvider.getWebSocketContainer();
         final MessageHandler handler = this;
         LOGGER.debug("initiating actions to websocket");
+        if (establishConnectionThread != null) {
+            establishConnectionThread.interrupt();
+        }
+        establishConnectionThread = null;
 
-        try {
-            websocketSession = client.connectToServer(new Endpoint()
-            {
-                @Override
-                public void onOpen(Session session, EndpointConfig config)
-                {
-                    session.addMessageHandler(handler);
-                }
+        establishConnectionThread = new Thread() {
```

Comment:
looks a bit complex here: I don't see the point to delegate the websocket connection to a thread. I would keep the old way to start the connection to the server and maybe use a locking mechanism to ensure only one connection attempt is made at the same time. I missed this yesterday with the code reformatting.